execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -256,32 +256,41 @@ where
         if !visited.insert(trie_key) {
             continue;
         }
-        let maybe_retrieved_trie: Option<Trie<K, V>> = store.get(txn, &trie_key)?;
+        let maybe_retrieved_trie_bytes: Option<Bytes> = store.get_raw(txn, &trie_key)?;
         // Perform an optional integrity check on the retrieved node.
         if check_integrity {
-            if let Some(trie_value) = &maybe_retrieved_trie {
-                let hash_of_trie_value = {
-                    let node_bytes = trie_value.to_bytes()?;
-                    Digest::hash(&node_bytes)
-                };
+            if let Some(trie_value_bytes) = &maybe_retrieved_trie_bytes {
+                let hash_of_trie_value = Digest::hash(trie_value_bytes);
                 if trie_key != hash_of_trie_value {
                     warn!(
-                        "Trie key {:?} has corrupted value {:?} (hash of value is {:?}); \
-                     adding to list of missing nodes",
-                        trie_key, trie_value, hash_of_trie_value
+                        "Trie key {:?} has corrupted value with hash {:?}; adding to list of missing nodes",
+                        trie_key, hash_of_trie_value,
                     );
                     missing_descendants.push(trie_key);
                     continue;
                 }
             }
         }
+
+        // Optimization: Don't look for descendants of leaves.
+        if let Some(bytes) = maybe_retrieved_trie_bytes.as_ref() {
+            // if this is a leaf
+            if !bytes.is_empty() && bytes[0] == 0u8 {
+                continue;
+            }
+        }
+
+        let maybe_retrieved_trie: Option<Trie<K, V>> = maybe_retrieved_trie_bytes
+            .map(bytesrepr::deserialize_from_slice)
+            .transpose()?;
```

Comment:
```suggestion
        match retrieved_trie: Option<Trie<K, V>> = maybe_retrieved_trie_bytes
            .map(bytesrepr::deserialize_from_slice)
            .transpose() {
            Ok(Some(retrieved_trie)) => retrieved_trie,
            Ok(None) => { missing_descendants.push(trie_key); continue }
            // Couldn't parse; treat as missing and continue
            Err(err) => {error!(err) ; missing_descendants.push(trie_key); continue }
        }
```

Again, if you decide to go with [my suggestion above][1] then you only have to treat the `Ok(retrieved_trie)` and `Err(error)` match cases...

[1]: https://github.com/casper-network/casper-node/pull/2680/files#r820254715

Reply:
See https://github.com/dwerner/casper-node/blob/218377c25745e707a9cb939fb0693d3c876c9107/execution_engine/src/storage/trie_store/operations/mod.rs#L291-L307